A built-in command to create a version 4 uuid

reneklootwijk opened the pull request:
Signed-off-by: Rene Klootwijk <rene.klootwijk@rencon.nl>

I included no test because the uuid command returns a random string which cannot be tested other than if it is a string.

reneklootwijk commented:
@tsandall : The reason I want to be able to create a uuid inside a policy is that I want to create an object according to a policy which also contains a uuid which is not provided as input.

Ad. 1: I do not understand what you mean. Why do you want to return the same uuid when calling the command multiple times from within the same policy? This would require keeping state outside of the policy. 

Ad. 2: I do not feel like changing the code to support a supplied seed just to check the code correctly formats the string seems overkill.

Ad. 3: Done

Ad. 4: Done (and in the mean time issued a PR for the google code to fix a bug ;-) )

If you agree with this I will create a new PR.

patrick-east commented:
> I do not understand what you mean. Why do you want to return the same uuid when calling the command multiple times from within the same policy? This would require keeping state outside of the policy.

OPA already does this caching/memoization of rules and builtin functions to avoid re-computing them. So there is already storage that happens in the background while the query is being evaluated. There are a couple of things to consider:

* Within a query/policy evaluation, OPA's evaluator does not provide guarantees that some expression will only be evaluated a single time. The policy author may write the function call a single time but under the hood it is fair game for OPA to call it >1 time for a given query evaluation. Where possible having idempotent behavior with the builtins is ideal. In this case where we are generating a random value in practice for the end result it is ideal to have 1 uuid generated per usage of `uuid()` in the policy versus 1 uuid generated per evaluation of the expression calling the builtin function.
* This, and maybe more to the point about providing a seed, helps not only for OPA unit testing, but for policy authors who use the uuid() helper function in their policies and allowing them to more easily write unit tests for the policies.

> Changing the code to support a supplied seed just to check the code correctly formats the string seems overkill.

If we just move this https://github.com/open-policy-agent/opa/blob/b39b34760e52a8db723423c89961328ff85b8612/runtime/runtime.go#L673-L682 over into like `internal/uuid/uuid.go` and share it for both of these it should be very easy to change the signature to take in a seed and use a new https://golang.org/pkg/math/rand/#Rand as the `io.reader` rather than the `rand.Reader` it uses right now.

> I included no test because the uuid command returns a random string which cannot be tested other than if it is a string.

We should, at a minimum, verify that the format we generate matches what we document (we don't actually document what format comes out, is it raw hex? with `-`'s?). You can also have tests that round trip from a policy, capture the result as a golang string, and validate that it is in fact a valid UUID.

reneklootwijk commented:
Sounds good to me, but you are talking changing some internals that I do not understand. Do you still expect something from me, or are you taking over?

patrick-east commented:
> Sounds good to me, but you are talking changing some internals that I do not understand. Do you still expect something from me, or are you taking over?

😃 I'm happy to help and provide guidance, but you're going to have to do the implementation.

The steps look like:
* Move that uuid helper I linked to into a new file like `internal/uuid/uuid.go`
* Rename it to use a capital letter(s) so its public, or maybe just call it `New()` or something (so in usage its `uuid.New()`)
* Add a `seed` int64 parameter for it, or even pass in a reader (almost identical to https://github.com/google/uuid/blob/master/version4.go#L34 at this point)
* Change from `io.ReadFull(rand.Reader, bs)` to be `rand.New(rand.NewSource(seed)).Read(bs)` (note, this switches from `crypto/rand` to `math/rand`)
   * Depending on if you used a seed parameter or a reader parameter**
* Change the two spots in `runtime.go` that called it to use the new signature https://github.com/open-policy-agent/opa/blob/master/runtime/runtime.go#L661-L671
  * If you need pointers for making a "good" seed check out https://stackoverflow.com/a/54491783/4789546
* Change the signature of the builtin function to take in a parameter
* Change the builtin implementation to use the seed and the newly shared helper
* Add tests for any new stuff

reneklootwijk commented:
I created the helper function for uuid, changed the runtime.go to use the helper, changed the builtin to use the helper, added the option to specify a seed and added a test. However I do not understand the caching of opa and how to use it to implement idempotency and how to prevent idempotency when different results are required.

patrick-east commented on ast/builtins.go:
```diff
@@ -982,6 +985,20 @@ var UnitsParseBytes = &Builtin{
+var UUID = &Builtin{
+	Name: "uuid.rfc4122",
+	Decl: types.NewFunction(
+		nil,
```
Please adjust the signature to take a `types.N` as the seed parameter.

patrick-east commented on internal/uuid/uuid.go:
```diff
@@ -0,0 +1,31 @@
+package uuid
```
nit: Please add the copyright header on new files (just copy it from one of the other opa source files)

patrick-east commented on internal/uuid/uuid.go:
```diff
@@ -0,0 +1,31 @@
+	"io"
+)
+
+var rander = rand.Reader
+
+// SetRand Use your own randomizer to support unit testing
+func SetRand(r io.Reader) {
+	if r == nil {
+		rander = rand.Reader
+		return
+	}
+
+	rander = r
+}
+
+// New Create a version 4 random UUID
+func New() (string, error) {
```
Instead of the global variable and `SetRand()` please change the signature of `New()` to take in the reader (almost exactly like https://github.com/google/uuid/blob/master/version4.go#L34)

Then where we call it in `runtime.go` pass in the `crypto/rand` `rand.Reader`, and for the builtin implementation use a new random reader with https://golang.org/pkg/math/rand/#New and a `Source` https://golang.org/pkg/math/rand/#NewSource configured to use the seed that was provided as the argument.

reneklootwijk replied:
I still do not understand why you want deterministic behaviour of a uuid function which sole purpose is to provide a truly random string (at least for version 4). I also wonder how you can provide a truly random seed as argument to the uuid function from within a policy. A better solution would be, imho, an optional seed where the crypto/rand Reader is used when no seed has been specified. However, I do not know how to implement optional parameters in Go (I am a novice in Go). I prefer to use the crypto/rand module over the math/rand because the quality of the random result will be significantly higher.

patrick-east replied:
Thank you for working with us on this @reneklootwijk I appreciate your patience on the issue.

> I still do not understand why you want deterministic behaviour of a uuid function which sole purpose is to provide a truly random string

There are two main reasons. First, and arguably the most (and maybe only) important one is that OPA does not evaluate these in an imperative top-to-bottom way. Just because the `uuid()` call is in an expression a single time in the policy it does not mean that it will be called a single time. That expression _could_ be evaluated any number of times by OPA while evaluating some top level query, to that end the builtin functions need should be implemented to be idempotent. The second reason is to make policies easier to unit test for policy authors. If they can predict the uuid for a unit test they can more easily compare expected and actual results for some query. Its the same reason why almost all languages/frameworks have options to seed "random" API's.

> I also wonder how you can provide a truly random seed as argument to the uuid function from within a policy.

> I prefer to use the crypto/rand module over the math/rand because the quality of the random result will be significantly higher.

Totally valid points. In my mind I was advocating for the argument to the builtin function to act as the actual random number seed, but it doesn't need to be the case to fulfill the requirements  I described above.

What I would propose is that we have the argument, but we don't call it a "seed". Internally use the builtin context cache https://github.com/open-policy-agent/opa/blob/5e6a1385595b7a09ee7f0184ac1168d2b7d48204/topdown/builtins.go#L34 which is provided as an argument to the builtin implementation if you use the newer registration functions rather than the deprecated `RegisterFunctionalBuiltinX` ones https://github.com/open-policy-agent/opa/blob/5e6a1385595b7a09ee7f0184ac1168d2b7d48204/topdown/builtins.go#L41-L52. 

So we use the parameter as the key, with a more optimally generated uuid as the value. Nothing prevents us from using `crypto/rand`'s reader to make a these UUID's. We then would just need to have a step early in the builtin to check the cache if it already has a uuid generated for the parameter, if it does return it, otherwise generate a new one (save it in the cache) and return that.

The new documented definition is more like:
| Built-in | Description |
| ------- |-------------|
| <span class="opa-keep-it-together">`output := uuid.rfc4122(str)`</span> | `output` is a `string` representing a version 4 UUID. For any given `str` the `output` will be consistent throughout a query evaluation. |

End result is that you generate the most random UUID available, and we can have the function be idempotent for a given parameter.

Thoughts?

tsandall replied:
Just to add onto what @patrick-east said, IMO there are two related but different things to consider here. Both have to do with determinism. I mentioned them very briefly in my original comment but didn't go into depth. Luckily @patrick-east went into some detail already so this might just be repetitive (sorry!)

Having a simple function that accepts no arguments and returns a random UUID may be fine at first glance, however, there are some interesting problems that it would create. For example, suppose you wrote the following:

```rego
package x

p = uuid()
```

And then somewhere else you said: `a = data.x.p; b = data.x.p`. In this case, I'd expect `a` and `b` to have the same value. However, if `uuid()` generated a random value on each invocation, that would not be true--this would be a fundamental change to the semantics. Unfortunately it's not even that simple because in this particular case, because of an optimization in the OPA evaluator for complete rules `a` and `b` WOULD have the same value because OPA memoizes complete rule values/assignments. So now the `uuid()` function does not even act the way we wanted at the outset.

The answer to this particular problem is to be explicit in the policy about when we want a new UUID. We can do this by simply extending the function to accept a parameter that identifies the unique value (@patrick-east described this nicely above.)

So that's the first thing. The other thing is that Rego is intended to be portable, meaning you can ship it around and execute it and get the same result out. This is useful for different kinds of policy enforcement (e.g., define the policy once and then apply it in multiple places like in a cluster, in a CI/CD pipeline, offline for audit purposes, etc.) but also for analysis and testing! Since the policies are deterministic (w/ the exception of a special class of built-in function) you can do things like replay decisions and be confident the answer is the same. This gets us back back to the point of being able to _seed_ the uuid function when evaluation is run. I don't have a strong opinion about whether the policy author should be able to seed or not but the evaluator definitely needs some way to do this otherwise what I said before is not possible. Note, we have the same problem with `time.now_ns()`  and `http.send()` though time and uuids are way easier to deal with.

reneklootwijk commented:
I added the suggested parameter and implemented idempotency using the caching feature. Because all those data types used in the builtin commands defined in the topdown package are a bit magic to me, I hoped I did it correctly. Some basic tests showed the idempotency seems to work.
